### ISSUE TITLE
Add js.Iterator according to ES6's iterator protocol

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#1141]
Doc: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/The_Iterator_protocol
Basically it boils down to:

``` scala
trait Iterator[+A] extends js.Object {
  def next(): Iterator.Entry[A] = native
}
object Iterator {
  trait Entry[+A] {
    val done: Boolean = native
    val value: A = native
  }
}
```

We should provide conversions to/from [`s.c.Iterator`](http://www.scala-lang.org/api/current/index.html#scala.collection.Iterator). Note that the API is not strictly the same, and therefore the `hasNext()` method of `s.c.Iterator` will have to call `next()` on `js.Iterator` (and cache its result).

This will allow us to interoperate smoothly with ES6's [generators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*).

**Comment by bblfish** [comment]
+1 useful for things like [Headers](https://fetch.spec.whatwg.org/#headers-class)
But it looks like this requires `@JSName(Symbol(Symbol.iterator))` to be able to access the `iterator` field of the Iterable.

```
@js.native
trait Iterable[+A] extends js.Object {
  @JSName("Symbol(Symbol.iterator)")
  def iterator(): Iterator[A] = js.native
}
```

Btw, is there a way to write native traits that do not extend `js.Object`? It looks like this trait would be usefully applied to a lot of other native classes, and there risks being a clash if a number of them extend `js.Object`

**Comment by sjrd** [comment]
This was implemented in #2784.
